gpjax/abstractions.py

```diff
@@ -125,42 +130,56 @@ def fit_batches(
     trainables: tp.Dict,
     train_data: Dataset,
     optax_optim,
+    seed: tp.Union[int, PRNGKey],
+    batch_size: int,
     n_iters: tp.Optional[int] = 100,
     log_rate: tp.Optional[int] = 10,
-) -> tp.Dict:
+) -> tp.Tuple[tp.Dict, f64["n_iters"]]:
     """Abstracted method for fitting a GP model with mini-batches respect to a supplied objective function.
     Optimisers used here should originate from Optax.
     Args:
         objective (tp.Callable): The objective function that we are optimising with respect to.
         params (dict): The parameters for which we would like to minimise our objective function with.
         trainables (dict): Boolean dictionary of same structure as 'params' that determines which parameters should be trained.
         train_data (Dataset): The training dataset.
-        opt_init (tp.Callable): The supplied optimiser's initialisation function.
-        opt_update (tp.Callable): Optimiser's update method.
-        get_params (tp.Callable): Return the current parameter state set from the optimiser.
+        optax_optim (GradientTransformation): The Optax optimiser that is to be used for learning a parameter set.
+        seed (int): The random seed for the mini-batch sampling.
+        batch_size(int): The batch_size.
         n_iters (int, optional): The number of optimisation steps to run. Defaults to 100.
         log_rate (int, optional): How frequently the objective function's value should be printed. Defaults to 10.
     Returns:
-        tp.Dict: An optimised set of parameters.
+        tp.Tuple[tp.Dict, f64["n_iters"]]: A tuple comprising optimised parameters and training history respectively.
     """
 
     opt_state = optax_optim.init(params)
-    next_batch = train_data.get_batcher()
+
+    prng = convert_seed(seed)
+
+    x, y, n = train_data.X, train_data.y, train_data.n
 
     def loss(params, batch):
         params = trainable_params(params, trainables)
         return objective(params, batch)
 
     @progress_bar_scan(n_iters, log_rate)
-    def step(params_opt_state, i):
-        params, opt_state = params_opt_state
-        batch = next_batch()
+    def step(carry, _):
+        params, opt_state, prng = carry
+
+        indicies = jr.choice(prng, n, (batch_size,), replace=True)
+
+        batch = Dataset(X=x[indicies], y=y[indicies])
+
         loss_val, loss_gradient = jax.value_and_grad(loss)(params, batch)
         updates, opt_state = optax_optim.update(loss_gradient, opt_state, params)
         params = optax.apply_updates(params, updates)
-        params_opt_state = params, opt_state
-        return params_opt_state, loss_val
 
-    (params, _), _ = jax.lax.scan(step, (params, opt_state), jnp.arange(n_iters))
+        prng, _ = jr.split(prng)
```

Comment:
Are you sure this is correct? Typically in Jax we'd split the key as `key, subkey = jr.split(key)` and then use subkey for stochastic ops. I believe this is because the return of `jr.split(key, 10) will first return a single key and the seond return is a list of 10 keys for use in a stochastic ops.

Reply:
Ooh I did not realise this. I will have to fix this. I think we might want to have the data batch loading code as a separate function so that we can unit test it.